messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/JGroupBroadcastGroupAdd.java

```diff
@@ -0,0 +1,161 @@
+import org.jboss.as.controller.OperationFailedException;
+import org.jboss.as.controller.OperationStepHandler;
+import org.jboss.as.controller.PathAddress;
+import org.jboss.as.controller.descriptions.ModelDescriptionConstants;
+import org.jboss.as.controller.operations.common.Util;
+import org.jboss.as.controller.registry.Resource;
+import org.jboss.dmr.ModelNode;
+import org.jboss.dmr.Property;
+import org.jboss.msc.service.ServiceController;
+import org.jboss.msc.service.ServiceName;
+import org.jboss.msc.service.ServiceRegistry;
+import org.wildfly.clustering.dispatcher.CommandDispatcherFactory;
+import org.wildfly.extension.messaging.activemq.broadcast.CommandDispatcherBroadcastEndpointFactory;
+import org.wildfly.extension.messaging.activemq.logging.MessagingLogger;
+
+/**
+ * Handler for adding a broadcast group.
+ *
+ * @author Brian Stansberry (c) 2011 Red Hat Inc.
+ */
+public class JGroupBroadcastGroupAdd extends AbstractAddStepHandler {
+
+    public static final JGroupBroadcastGroupAdd INSTANCE = new JGroupBroadcastGroupAdd(true);
+    public static final JGroupBroadcastGroupAdd LEGACY_INSTANCE = new JGroupBroadcastGroupAdd(false);
+
+    private final boolean needLegacyCall;
+    private JGroupBroadcastGroupAdd(boolean needLegacyCall) {
+        super(JGroupBroadcastGroupDefinition.ATTRIBUTES);
+        this.needLegacyCall= needLegacyCall;
+    }
+
+    @Override
+    public void execute(OperationContext context, ModelNode operation) throws OperationFailedException {
+        CommonAttributes.renameChannelToCluster(operation);
+        if (operation.hasDefined(JGROUPS_CLUSTER.getName())) {
+            if (operation.hasDefined(JGROUPS_CHANNEL_FACTORY.getName()) && !operation.hasDefined(JGROUPS_CHANNEL.getName())) {
+                // Handle legacy behavior
+                String channel = operation.get(JGROUPS_CLUSTER.getName()).asString();
+                operation.get(JGROUPS_CHANNEL.getName()).set(channel);
+
+                PathAddress channelAddress = context.getCurrentAddress().getParent().getParent().getParent().append(ModelDescriptionConstants.SUBSYSTEM, "jgroups").append("channel", channel);
+                ModelNode addChannelOperation = Util.createAddOperation(channelAddress);
+                addChannelOperation.get("stack").set(operation.get(JGROUPS_CHANNEL_FACTORY.getName()));
+                // Fabricate a channel resource if it is missing
+                context.addStep(addChannelOperation, AddIfAbsentStepHandler.INSTANCE, OperationContext.Stage.MODEL);
+            }
+        }
+        if(needLegacyCall) {
+            PathAddress target = context.getCurrentAddress().getParent().append(CommonAttributes.BROADCAST_GROUP, context.getCurrentAddressValue());
+            ModelNode op = operation.clone();
+            op.get(OP_ADDR).set(target.toModelNode());
+            context.addStep(op, BroadcastGroupAdd.LEGACY_INSTANCE, OperationContext.Stage.MODEL, true);
+        }
```

Comment:
Why is this needed?
If a user issues an add operation for the new resource, we should not need to create any legacy resources.  These would only need to exist if the user relies on commands from the old model version (in which case the add operation would have been issues for the legacy resource).

Reply:
What if the user mixes old and new commands. We need a 'legacy' resource just to handle the operations

Reply:
You're right.  We need to handle mixed domains.